bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java

```diff
@@ -0,0 +1,397 @@
+     */
+    public SunspecDiscoveryProcess(ModbusEndpointThingHandler handler, ModbusDiscoveryListener listener)
+            throws EndpointNotInitializedException {
+        this.handler = handler;
+
+        ModbusSlaveEndpoint endpoint = this.handler.asSlaveEndpoint();
+        if (endpoint != null) {
+            this.endpoint = endpoint;
+        } else {
+            throw new EndpointNotInitializedException();
+        }
+        slaveId = handler.getSlaveId();
+        this.listener = listener;
+        commonBlockParser = new CommonModelParser();
+        possibleAddresses = new CopyOnWriteArrayList<>();
+        // Preferred and alternate base registers
+        // @see SunSpec Information Model Overview
+        possibleAddresses.add(40000);
+        possibleAddresses.add(50000);
+        possibleAddresses.add(0);
+    }
+
+    /**
+     * Start model detection
+     *
+     * @param uid the thing type to look for
+     * @throws EndpointNotInitializedException
+     */
+    public void detectModel() {
+
+        if (possibleAddresses.size() < 1) {
+            parsingFinished();
+            return;
+        }
+        // Try the next address from the possibles
+        baseAddress = possibleAddresses.get(0);
+        logger.trace("Beginning scan for SunSpec device at address {}", baseAddress);
+
+        BasicModbusReadRequestBlueprint request = new BasicModbusReadRequestBlueprint(slaveId,
+                ModbusReadFunctionCode.READ_MULTIPLE_REGISTERS, baseAddress, // Start address
+                SUNSPEC_ID_SIZE, // number or words to return
+                maxTries);
+
+        PollTask task = new BasicPollTaskImpl(endpoint, request, new ModbusReadCallback() {
+
+            @Override
+            public void onRegisters(ModbusReadRequestBlueprint request, ModbusRegisterArray registers) {
+                headerReceived(registers);
+            }
+
+            @Override
+            public void onError(ModbusReadRequestBlueprint request, Exception error) {
+                handleError(error);
+            }
+
+            @Override
+            public void onBits(@Nullable ModbusReadRequestBlueprint request, @Nullable BitArray bits) {
+                // don't care, we don't expect this result
+            }
+        });
+
+        handler.getManagerRef().get().submitOneTimePoll(task);
+    }
+
+    /**
+     * We received the first two words, that should equal to SunS
+     */
+    private void headerReceived(ModbusRegisterArray registers) {
+        logger.trace("Received response from device {}", registers.toString());
+
+        Optional<DecimalType> id = ModbusBitUtilities.extractStateFromRegisters(registers, 0, ValueType.UINT32);
+
+        if (!id.isPresent() || id.get().longValue() != SUNSPEC_ID) {
+            logger.debug("Could not find SunSpec DID at address {}, received: {}, expected: {}", baseAddress, id,
+                    SUNSPEC_ID);
+            possibleAddresses.remove(0);
```

Comment:
If you poll within detectModel() you can remove this

Reply:
thanks, removed.